flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/pattern/conditions/RichAndCondition.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.cep.pattern.conditions;
+
+import org.apache.flink.api.common.functions.RuntimeContext;
+import org.apache.flink.api.common.functions.util.FunctionUtils;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.util.Preconditions;
+
+/**
+ * A {@link RichIterativeCondition condition} which combines two conditions with a logical
+ * {@code AND} and returns {@code true} if both are {@code true}.
+ *
+ * @param <T> Type of the element to filter
+ */
+public class RichAndCondition<T> extends RichIterativeCondition<T> {
```

Comment:
How about we introduce `RichCompositeIterativeCondition` that will get list of conditions in the ctor and will handle the `open/close/setRuntimeContext` methods in one place?

Reply:
I agree to deal with the open and close methods in a way that adds abstract classes, and RichAnd is very similar to RichOr, but how do we deal with the difference between RichNot and RichAnd/RichOr?

Reply:
I though about sth like this:

    abstract class RichCompositeCondition<T> extends RichIterativeCondition<T> {
    
    	private IterativeCondition<T>[] conditions;
    
    	RichCompositeCondition(IterativeCondition<T>... conditions) {
    		this.conditions = conditions;
    	}
    
    	@Override
    	public void setRuntimeContext(RuntimeContext runtimeContext) {
    		super.setRuntimeContext(runtimeContext);
    
    		for (IterativeCondition<T> condition : conditions) {
    			FunctionUtils.setFunctionRuntimeContext(condition, runtimeContext);
    		}
    	}
    
    	@Override
    	public void open(Configuration parameters) throws Exception {
    		super.open(parameters);
    
    		for (IterativeCondition<T> condition : conditions) {
    			openFunction(condition, parameters);
    		}
    	}
    
    	@Override
    	public void close() throws Exception {
    		super.close();
    
    		for (IterativeCondition<T> condition : conditions) {
    			closeFunction(condition);
    		}
    	}
    }

    public class RichAndCondition<T> extends RichCompositeCondition<T> {
    
    	private static final long serialVersionUID = 1L;
    
    	private final IterativeCondition<T> left;
    	private final IterativeCondition<T> right;
    
    	public RichAndCondition(final IterativeCondition<T> left, final IterativeCondition<T> right) {
    		super(left, right);
    		this.left = Preconditions.checkNotNull(left, "The condition cannot be null.");
    		this.right = Preconditions.checkNotNull(right, "The condition cannot be null.");
    
    	}
    
    	@Override
    	public boolean filter(T value, Context<T> ctx) throws Exception {
    		return left.filter(value, ctx) && right.filter(value, ctx);
    	}
    
    	/**
    	 * @return One of the {@link IterativeCondition conditions} combined in this condition.
    	 */
    	public IterativeCondition<T> getLeft() {
    		return left;
    	}
    
    	/**
    	 * @return One of the {@link IterativeCondition conditions} combined in this condition.
    	 */
    	public IterativeCondition<T> getRight() {
    		return right;
    	}
    }

    public class RichNotCondition<T> extends RichCompositeCondition<T> {
    
    	private static final long serialVersionUID = 1L;
    
    	private final IterativeCondition<T> original;
    
    	public RichNotCondition(final IterativeCondition<T> original) {
    		super(original);
    		this.original = Preconditions.checkNotNull(original);
    	}
    
    	@Override
    	public boolean filter(T value, Context<T> ctx) throws Exception {
    		return original != null && !original.filter(value, ctx);
    	}
    }

Reply:
Variable parameters are a good idea! Does `RichNot ` need check null of Condition ?